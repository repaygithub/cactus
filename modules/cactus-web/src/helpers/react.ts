import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
