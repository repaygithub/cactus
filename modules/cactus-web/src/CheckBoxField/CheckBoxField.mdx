---
name: CheckBoxField
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import CheckBoxField from './CheckBoxField'
import cactusTheme from '@repay/cactus-theme'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'

# CheckBoxField

The `CheckBoxField` is a combination of `CheckBox` and `Label`, along with some accessibility features that are taken care of for you.

### Try it out

export const code = `<CheckBoxField name="checkboxfield" label="Is Checked"/>`

<LiveProvider code={code} scope={{ CheckBoxField }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- Use meaningful labels for each `CheckBoxField` so that they are accessible.
- If you don't provide an ID, a unique ID will be generated for you.

## Basic usage

```jsx
import React, { useState } from 'react'
import { Button, CheckBoxField } from '@repay/cactus-web'

export default function Form() {
  const [values, setValues] = useState({ checked: false })
  const handleChange = useCallback(
    (name, value) => {
      setValues(state => setValues({ ...state, [name]: value }))
    },
    [setValues]
  )
  const onSubmit = useCallback(
    event => {
      event.preventDefault()
      const data = { ...values }
      // send data to api
    },
    [values]
  )

  return (
    <form onSubmit={onSubmit}>
      <CheckBoxField
        name="checkboxfield"
        label="Is Checked"
        onChange={handleChange}
        value={values.checked}
      />
      <Button type="submit" variant="action">
        Submit
      </Button>
    </form>
  )
}
```

## Properties

<PropsTable of={CheckBoxField} />
