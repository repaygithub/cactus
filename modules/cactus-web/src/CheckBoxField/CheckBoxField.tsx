import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

import CheckBox, { CheckBoxProps } from '../CheckBox/CheckBox'
import FieldWrapper from '../FieldWrapper/FieldWrapper'
import { extractFieldStyleProps } from '../helpers/omit'
import { FlexItemProps } from '../helpers/styled'
import useId from '../helpers/useId'
import Label, { LabelProps } from '../Label/Label'

export interface CheckBoxFieldProps extends CheckBoxProps, FlexItemProps {
  label: React.ReactNode
  labelProps?: Omit<LabelProps, 'children' | 'htmlFor'>
  id?: string
  name?: string
  disabled?: boolean
}

const CheckBoxFieldBase = React.forwardRef<HTMLInputElement, CheckBoxFieldProps>((props, ref) => {
  const { label, labelProps, id, name, ...checkboxProps } = props
  const styleProps = extractFieldStyleProps(checkboxProps)
  const checkboxId = useId(id, name)

  return (
    <FieldWrapper {...styleProps} $gap={3}>
      <CheckBox {...checkboxProps} ref={ref} id={checkboxId} name={name} />
      <Label {...labelProps} htmlFor={checkboxId}>
        {label}
      </Label>
    </FieldWrapper>
  )
})

export const CheckBoxField = styled(CheckBoxFieldBase)`
  ${Label} {
    cursor: ${(p): string => (p.disabled ? 'not-allowed' : 'pointer')};
    padding-left: 8px;
    color: ${(p) => p.disabled && p.theme.colors.mediumGray};
  }
`

CheckBoxField.propTypes = {
  label: PropTypes.node.isRequired,
  labelProps: PropTypes.object,
  id: PropTypes.string,
  name: PropTypes.string,
  disabled: PropTypes.bool,
}

export default CheckBoxField
