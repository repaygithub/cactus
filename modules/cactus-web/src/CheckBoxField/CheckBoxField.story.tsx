import React from 'react'
import { storiesOf } from '@storybook/react'
import { boolean, text } from '@storybook/addon-knobs'
import CheckBoxField from './CheckBoxField'

const checkBoxFieldStories = storiesOf('CheckBoxField', module)

checkBoxFieldStories.add('Basic Usage', () => (
  <CheckBoxField
    name={text('name', 'CheckBoxFormField')}
    id={text('id', 'checkbox-1')}
    label={text('label', 'A Label')}
    disabled={boolean('disabled', false)}
  />
))

checkBoxFieldStories.add('Multiple CheckBox Fields', () => (
  <div>
    <CheckBoxField name="CheckBoxFormField" label="Label 1" />
    <CheckBoxField name="CheckBoxFormField" label="Label 2" />
    <CheckBoxField name="CheckBoxFormField" label="Label 3" />
  </div>
))
