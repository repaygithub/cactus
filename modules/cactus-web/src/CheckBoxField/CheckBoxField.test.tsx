import * as React from 'react'
import { cleanup, render, fireEvent, queryByAttribute } from 'react-testing-library'
import userEvent from 'user-event'
import CheckBoxField from './CheckBoxField'
import cactusTheme from '@repay/cactus-theme'
import { ThemeProvider } from 'styled-components'

afterEach(cleanup)

describe('component: CheckBoxField', () => {
  test('should render a checkbox field', () => {
    const checkboxField = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="SoIA" id="my-id" />
      </ThemeProvider>
    )

    expect(checkboxField.asFragment()).toMatchSnapshot()
  })

  test('should render a disabled checkbox field', () => {
    const checkboxField = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="NMNL" id="my-id" disabled />
      </ThemeProvider>
    )

    expect(checkboxField.asFragment()).toMatchSnapshot()
  })

  test('should generate unique id when one is not provided', () => {
    const { container, getByText } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="Scoreboard" />
      </ThemeProvider>
    )

    const getById = queryByAttribute.bind(null, 'id')

    const labelElement = getByText('Scoreboard') as HTMLLabelElement
    expect(labelElement.htmlFor).toContain('rand')
    expect(getById(container, labelElement.htmlFor)).not.toBeNull()
  })

  test('should trigger onChange event', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="Katastro" onChange={onChange} data-testid="will-check" />
      </ThemeProvider>
    )

    fireEvent.click(getByTestId('will-check'))
    expect(onChange).toHaveBeenCalled()
  })

  test('should trigger onFocus event', () => {
    const onFocus = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="Strange Nights" onFocus={onFocus} data-testid="will-check" />
      </ThemeProvider>
    )

    fireEvent.focus(getByTestId('will-check'))
    expect(onFocus).toHaveBeenCalled()
  })

  test('should trigger onBlur event', () => {
    const onBlur = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="Washed." onBlur={onBlur} data-testid="will-check" />
      </ThemeProvider>
    )

    fireEvent.blur(getByTestId('will-check'))
    expect(onBlur).toHaveBeenCalled()
  })

  test('should not trigger onChange event', () => {
    const onChange = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField label="Flow" onChange={onChange} data-testid="will-not-check" disabled />
      </ThemeProvider>
    )

    userEvent.click(getByTestId('will-not-check'))
    expect(onChange).not.toHaveBeenCalled()
  })

  test('should not trigger onFocus event', () => {
    const onFocus = jest.fn()
    const { getByTestId } = render(
      <ThemeProvider theme={cactusTheme}>
        <CheckBoxField
          label="Not For Sale"
          onFocus={onFocus}
          data-testid="will-not-check"
          disabled
        />
      </ThemeProvider>
    )

    userEvent.click(getByTestId('will-not-check'))
    expect(onFocus).not.toHaveBeenCalled()
  })
})
