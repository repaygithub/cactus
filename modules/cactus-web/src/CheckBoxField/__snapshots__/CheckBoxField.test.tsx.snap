// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: CheckBoxField should render a checkbox field 1`] = `
<DocumentFragment>
  .c7 {
  vertical-align: middle;
}

.c5 {
  opacity: 0;
  border: 0;
  width: 0;
  height: 0;
  position: absolute;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border: 1px solid;
  border-color: hsl(200,10%,20%);
  background: none;
  border-radius: 1px;
}

.c6 svg {
  margin: 1px;
}

.c6 svg {
  visibility: hidden;
  display: block;
  color: hsl(0,0%,100%);
  width: 12px;
  height: 12px;
}

.c4 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  width: 16px;
  height: 16px;
  line-height: 16px;
  cursor: pointer;
}

.c4 input:checked ~ span {
  border-color: hsl(200,96%,35%);
  background-color: hsl(200,96%,35%);
}

.c4 input:checked ~ span svg {
  visibility: visible;
}

.c4 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c1 + .c0 {
  margin-top: 16px;
}

.c9 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c3 + .c2 {
  margin-top: 8px;
}

.c3 .c8 {
  cursor: pointer;
  padding-left: 8px;
}

<div
    class="c0 c1 c2 c3"
  >
    <label
      class="c4"
      for="my-id"
    >
      <input
        class="c5"
        id="my-id"
        name="checkbox-test"
        type="checkbox"
      />
      <span
        aria-hidden="true"
        class="c6"
      >
        <svg
          class="c7"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
          />
        </svg>
      </span>
    </label>
    <label
      class="c8 c9"
      for="my-id"
    >
      SoIA
    </label>
  </div>
</DocumentFragment>
`;

exports[`component: CheckBoxField should render a disabled checkbox field 1`] = `
<DocumentFragment>
  .c7 {
  vertical-align: middle;
}

.c5 {
  opacity: 0;
  border: 0;
  width: 0;
  height: 0;
  position: absolute;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border: 1px solid;
  border-color: hsl(0,0%,90%);
  background: hsl(0,0%,90%);
  border-radius: 1px;
}

.c6 svg {
  margin: 1px;
}

.c6 svg {
  visibility: hidden;
  display: block;
  color: hsl(0,0%,100%);
  width: 12px;
  height: 12px;
}

.c4 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  width: 16px;
  height: 16px;
  line-height: 16px;
  cursor: cursor;
}

.c4 input:checked ~ span svg {
  visibility: visible;
}

.c4 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c1 + .c0 {
  margin-top: 16px;
}

.c9 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c10 + .c2 {
  margin-top: 8px;
}

.c10 .c8 {
  cursor: pointer;
  padding-left: 8px;
}

.c3 + .c2 {
  margin-top: 8px;
}

.c3 .c8 {
  cursor: not-allowed;
  padding-left: 8px;
}

<div
    class="c0 c1 c2 c3"
  >
    <label
      class="c4"
      for="my-id"
    >
      <input
        class="c5"
        disabled=""
        id="my-id"
        name="checkbox-test"
        type="checkbox"
      />
      <span
        aria-hidden="true"
        class="c6"
        disabled=""
      >
        <svg
          class="c7"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
          />
        </svg>
      </span>
    </label>
    <label
      class="c8 c9"
      for="my-id"
    >
      NMNL
    </label>
  </div>
</DocumentFragment>
`;

exports[`component: CheckBoxField should support margin space props 1`] = `
<DocumentFragment>
  .c7 {
  vertical-align: middle;
}

.c5 {
  opacity: 0;
  border: 0;
  width: 0;
  height: 0;
  position: absolute;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  width: 16px;
  height: 16px;
  border: 1px solid;
  border-color: hsl(200,10%,20%);
  background: none;
  border-radius: 1px;
}

.c6 svg {
  margin: 1px;
}

.c6 svg {
  visibility: hidden;
  display: block;
  color: hsl(0,0%,100%);
  width: 12px;
  height: 12px;
}

.c4 {
  position: relative;
  display: inline-block;
  vertical-align: -1px;
  width: 16px;
  height: 16px;
  line-height: 16px;
  cursor: pointer;
}

.c4 input:checked ~ span {
  border-color: hsl(200,96%,35%);
  background-color: hsl(200,96%,35%);
}

.c4 input:checked ~ span svg {
  visibility: visible;
}

.c4 input:focus ~ span {
  box-shadow: 0px 3px 8px hsla(200,96%,35%,0.3);
}

.c1 + .c0 {
  margin-top: 16px;
}

.c9 {
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(200,10%,20%);
}

.c10 + .c2 {
  margin-top: 8px;
}

.c10 .c8 {
  cursor: pointer;
  padding-left: 8px;
}

.c11 + .c2 {
  margin-top: 8px;
}

.c11 .c8 {
  cursor: not-allowed;
  padding-left: 8px;
}

.c3 {
  margin-right: 8px;
}

.c3 + .c2 {
  margin-top: 8px;
}

.c3 .c8 {
  cursor: pointer;
  padding-left: 8px;
}

<div
    class="c0 c1 c2 c3"
  >
    <label
      class="c4"
      for="not-random"
    >
      <input
        class="c5"
        id="not-random"
        name="space_props"
        type="checkbox"
      />
      <span
        aria-hidden="true"
        class="c6"
      >
        <svg
          class="c7"
          fill="currentcolor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M9.81384674,19.7396292 C9.5249491,19.7396292 9.24415138,19.6325301 9.02725315,19.4381317 L3.38789908,14.3504731 C2.906403,13.9166766 2.86770331,13.1741827 3.30239977,12.6926866 C3.73619624,12.2120905 4.47869019,12.1724908 4.96018627,12.6071873 L9.67974783,16.8641526 L18.9127726,5.4369457 C19.3204693,4.93204981 20.0593633,4.85285046 20.5642592,5.26144713 C21.0682551,5.66914381 21.1465544,6.40803779 20.7388578,6.91293368 L10.7273393,19.3040328 C10.524841,19.5542307 10.2269434,19.7108295 9.90654599,19.7369292 C9.87504624,19.7387292 9.84444649,19.7396292 9.81384674,19.7396292"
          />
        </svg>
      </span>
    </label>
    <label
      class="c8 c9"
      for="not-random"
    >
      space props
    </label>
  </div>
</DocumentFragment>
`;
